### PR TITLE
VX-134: osin in-memory storage

### PR DESCRIPTION
**Pull request by yarencheng**

### CODE REVIEW

File: src/osin_storages/memory.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
+package osin_storages
```
Comment (c9s) [inline src/osin_storages/memory.go]: So we have to implement a new one? Is there a builtin storage? 

Also , please avoid using underscore in the package name

**Review by c9s — CHANGES_REQUESTED** [review]

File: src/osin_storages/memory.go (hunk 1 above)
Comment (yarencheng) [reply src/osin_storages/memory.go]: It is copied from osin/example https://github.com/RangelReale/osin/blob/master/example/teststorage.go

A mongo version already exist https://github.com/linkernetworks/oauth/blob/develop/src/app/storage/storage.go

I think in-memory version is more convenient.

**Comment by hwchiu** [comment]
需要針對其他的read/write功能寫測試嗎?

**Comment by yarencheng** [comment]
以後會加，但是因為目前 read/write 都太簡單了，暫時不想花時間寫相關的測試，而且還不確定以後會用in-memory 還是 mongodb

**Review by c9s — APPROVED** [review]

File: src/osinstorage/memory_test.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
+package osinstorage
+
+import (
+	"testing"
+
+	"github.com/RangelReale/osin"
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_Clone(t *testing.T) {
```
Comment (c9s) [inline src/osinstorage/memory_test.go]: Please Do not put undersore in the name

File: src/osinstorage/memory.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+	clientsMutex   sync.Mutex
+	authorize      map[string]osin.AuthorizeData
+	authorizeMutex sync.Mutex
+	access         map[string]osin.AccessData
+	accessMutex    sync.Mutex
+	refresh        map[string]string
+	refreshMutex   sync.Mutex
+}
+
+func NewMemory(clients ...osin.DefaultClient) *MemoryStorage {
```
Comment (c9s) [inline src/osinstorage/memory.go]: When you use NewXXX the XXX must follow the struct name

**Review by c9s — CHANGES_REQUESTED** [review]

**Comment by c9s** [comment]
Production 一定是用 mongodb, memory 只是用來測試

**Review by hwchiu — APPROVED** [review]

**Review by c9s — APPROVED** [review]

**Pull request merged**